python/cudf/cudf/io/parquet.py

```diff
@@ -561,6 +565,118 @@ def read_parquet(
+
+    # Return final cudf.DataFrame
+    return df.reset_index(drop=True) if reset_index else df
+
+
+def _apply_post_filters(df, filters):
+    # Apply DNF filters to an in-memory DataFrame
+    #
+    # Disjunctive normal form (DNF) means that the inner-most
+    # tuple describes a single column predicate. These inner
+    # predicates are combined with an AND conjunction into a
+    # larger predicate. The outer-most list then combines all
+    # of the combined filters with an OR disjunction.
+
+    if not filters:
+        # No filters to apply
+        return df
+
+    _comparisons = {
+        "==": operator.eq,
+        "!=": operator.ne,
+        "<": operator.lt,
+        "<=": operator.le,
+        ">": operator.gt,
+        ">=": operator.ge,
+    }
+
+    try:
+        # Disjunction loop
+        #
+        # All elements of `disjunctions` shall be combined with
+        # an `OR` disjunction (operator.or_)
+        disjunctions = []
+        other = filters.copy()
+        while other:
+
+            # Conjunction loop
+            #
+            # All elements of `conjunctions` shall be combined with
+            # an `AND` conjunction (operator.and_)
+            conjunctions = []
+            comparisons, *other = (
+                other if isinstance(other[0], list) else [other]
+            )
+            for (column, op, value) in comparisons:
+
+                # Inner comparison loop
+                #
+                # `op` is expected to be the string representation
+                # of a comparison operator (e.g. "==")
+                if op == "in":
+                    # Special case: "in"
+                    if not isinstance(value, (list, set, tuple)):
+                        raise TypeError(
+                            "Value of 'in' filter must be a "
+                            "list, set, or tuple."
+                        )
+                    if len(value) == 1:
+                        conjunctions.append(operator.eq(df[column], value[0]))
+                    else:
+                        conjunctions.append(
+                            operator.or_(
+                                *[operator.eq(df[column], v) for v in value]
+                            )
+                        )
+                elif op in ("is", "is not"):
+                    # Special case: "is" or "is not"
+                    if value not in (np.nan, None):
+                        raise TypeError(
+                            "Value of 'is' or 'is not' filter "
+                            "must be np.nan or None."
+                        )
+                    conjunctions.append(
+                        df[column].isna() if op == "is" else ~df[column].isna()
+                    )
+                else:
+                    # Conventional comparison operator
+                    conjunctions.append(_comparisons[op](df[column], value))
+
+            disjunctions.append(
+                operator.and_(*conjunctions)
```

Comment:
`and_` has type `bool -> bool -> bool` so this only works if you have exactly two conjunctions.

You want (handles length-1 conjunctions as well):

```
disjunctions.append(functools.reduce(operator.and_, conjunctions))
```

Reply:
Ah right - I did completely forget that `operator.and_`/`or_` are binops